[NVPTX] support packed f32 instructions for sm_100+

Prince781 opened the pull request:
This adds support for lowering `fadd`, `fsub`, `fmul`, and `fma` to sm_100+ packed-f32 instructions[^1] (e.g. `add.rn.f32x2 Int64Reg, Int64Reg`). <s>Rather than legalizing `v2f32`, we handle these four instructions ad hoc, so that codegen remains the same unless these instructions are present. We also introduce some DAGCombiner rules to simplify bitwise packing/unpacking to use `mov`, and to reduce redundant `mov`s.</s>

In this PR I didn't implement support for alternative rounding modes, as that was lower priority. If there's sufficient demand, I can add that to this PR. Otherwise we can leave that for later.

[^1]: Introduced in PTX 8.6: https://docs.nvidia.com/cuda/parallel-thread-execution/#changes-in-ptx-isa-version-8-6

AlexMaclean commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -818,6 +821,10 @@ def step_vector : SDNode<"ISD::STEP_VECTOR", SDTypeProfile<1, 1,
 def scalar_to_vector : SDNode<"ISD::SCALAR_TO_VECTOR", SDTypeProfile<1, 1, []>,
                               []>;
 
+def build_pair : SDNode<"ISD::BUILD_PAIR", SDTypeProfile<1, 2,
```
https://github.com/llvm/llvm-project/blob/cf5947be13e7af67219379e07bc0128f1f1e7f88/llvm/include/llvm/CodeGen/ISDOpcodes.h#L233-L236

BUILD_PAIR should not survive past legalization.

Prince781 replied:
Here I'm using it to replace `BUILD_VECTOR` as I'll be lowering it to a packing `mov.b64` later. It's very useful compared to defining a new opcode in NVPTXISD.

AlexMaclean reviewed (commented):
> Rather than legalizing v2f32, we handle these four instructions ad hoc, so that codegen remains the same unless these instructions are present.

Supporting `v2f32` (similar to how we support `v2f16` for example) would be a cleaner and more extensible way to implement this change.

Prince781 commented:
@AlexMaclean what led me to this implementation (and I did try it the other way) is that v2f16 and v2bf16 are supported by many kinds of instructions, so it makes more sense to legalize these types than v2f32. My concern is whether this feature should change code that uses f32 vectors but avoids these operations. Legalizing this type requires me to change some things in how we lower instructions, like loads and stores (for example, if we don't want `ld.v2.f32` to always become `ld.b64`), override a few things in TLI, etc. 

See the test cases for more examples. If this is not a concern, then I can implement it the other way.

Prince781 commented:
Okay, so I'm in the process of reworking this PR to legalize `v2f32`. If there's interest in the current approach, feel free to review. Otherwise, please wait until I update this.

Artem-B commented on llvm/test/CodeGen/NVPTX/f32x2-instructions.ll:
```diff
@@ -0,0 +1,2665 @@
+; CHECK-O3-NEXT:    add.rn.f32x2 %rd4, %rd2, %rd3;
+; CHECK-O3-NEXT:    st.param.b64 [func_retval0], %rd4;
+; CHECK-O3-NEXT:    ret;
+  %r = fadd <2 x float> %a, %b
```
I think that the patch may not be taking the right approach here.

Whether a particular instruction is supported or not should be a property of the target, and should not depend on the optimization level.

Optimizations may control whether it's better to use scalars of v2f32, but once we get to the lowering phase, if IR says v2f32 and we can lower it to a v2f32 operation, we should do that. If the target does not support it, *then* legalization kicks in and we split it

Prince781 replied:
> Whether a particular instruction is supported or not should be a property of the target, and should not depend on the optimization level.

Agree in general, but shouldn't we want to check that optimizations may not introduce new instructions that are unhandled? For example, this is currently broken on `-O3`:

https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/NVPTX/f16x2-instructions.ll#L76-L94

```
% llc < test-extract-i-f16x2.ll -mtriple=nvptx64-nvidia-cuda -O3 -verify-machineinstrs                                                                              :( 134 25-02-13 - 13:42:27
//
// Generated by LLVM NVPTX Back-End
//

.version 6.0
.target sm_30
.address_size 64


# After Instruction Selection
# Machine code for function test_extract_i: IsSSA, TracksLiveness

bb.0 (%ir-block.0):
  %0:int64regs = LD_i64_avar 0, 0, 101, 1, 0, 64, &test_extract_i_param_1 :: (dereferenceable invariant load (s64) from `ptr addrspace(101) null`, addrspace 101)
  %1:int64regs = ANDb64ri killed %0:int64regs, 1
  %2:int64regs = SHLi64ri killed %1:int64regs, 1
  %3:int64regs = ADDi64rr &test_extract_i_param_0, killed %2:int64regs
  %4:int16regs = LD_i16_areg_64 0, 0, 0, 1, 3, 16, killed %3:int64regs :: (dereferenceable invariant load (s16), addrspace 101)
  StoreRetvalI16 killed %4:int16regs, 0 :: (store (s16), align 1)
  Return

# End machine code for function test_extract_i.

*** Bad machine code: Expected a register operand. ***
- function:    test_extract_i
- basic block: %bb.0  (0x459ca9f0)
- instruction: %3:int64regs = ADDi64rr &test_extract_i_param_0, killed %2:int64regs
- operand 1:   &test_extract_i_param_0
LLVM ERROR: Found 1 machine code errors.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: llc -mtriple=nvptx64-nvidia-cuda -O3 -verify-machineinstrs
1.	Running pass 'Function Pass Manager' on module '<stdin>'.
2.	Running pass 'Verify generated machine code' on function '@test_extract_i'
 #0 0x0000000007e11b9c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/pferro/repos/llvm-project/llvm/lib/Support/Unix/Signals.inc:804:22
 #1 0x0000000007e11fda PrintStackTraceSignalHandler(void*) /home/pferro/repos/llvm-project/llvm/lib/Support/Unix/Signals.inc:880:1
 #2 0x0000000007e0f792 llvm::sys::RunSignalHandlers() /home/pferro/repos/llvm-project/llvm/lib/Support/Signals.cpp:105:20
 #3 0x0000000007e11546 SignalHandler(int, siginfo_t*, void*) /home/pferro/repos/llvm-project/llvm/lib/Support/Unix/Signals.inc:418:13
 #4 0x00007f0dd0140090 __restore_rt (/lib64/libc.so.6+0x1a090)
 #5 0x00007f0dd01990f4 __pthread_kill_implementation (/lib64/libc.so.6+0x730f4)
 #6 0x00007f0dd013ffde gsignal (/lib64/libc.so.6+0x19fde)
 #7 0x00007f0dd0127942 abort (/lib64/libc.so.6+0x1942)
 #8 0x0000000007d39fd2 llvm::report_fatal_error(llvm::Twine const&, bool) /home/pferro/repos/llvm-project/llvm/lib/Support/ErrorHandling.cpp:126:9
 #9 0x000000000682df5b (anonymous namespace)::MachineVerifier::ReportedErrors::~ReportedErrors() /home/pferro/repos/llvm-project/llvm/lib/CodeGen/MachineVerifier.cpp:262:33
#10 0x000000000682e0dc (anonymous namespace)::MachineVerifier::~MachineVerifier() /home/pferro/repos/llvm-project/llvm/lib/CodeGen/MachineVerifier.cpp:102:8
#11 0x000000000682e214 (anonymous namespace)::MachineVerifierLegacyPass::runOnMachineFunction(llvm::MachineFunction&) /home/pferro/repos/llvm-project/llvm/lib/CodeGen/MachineVerifier.cpp:391:12
#12 0x000000000667682c llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /home/pferro/repos/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:94:33
#13 0x0000000006ee647f llvm::FPPassManager::runOnFunction(llvm::Function&) /home/pferro/repos/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1406:20
#14 0x0000000006ee66e1 llvm::FPPassManager::runOnModule(llvm::Module&) /home/pferro/repos/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1452:13
#15 0x0000000006ee6aee (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /home/pferro/repos/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1521:20
#16 0x0000000006ee21b1 llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/pferro/repos/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:539:13
#17 0x0000000006ee7319 llvm::legacy::PassManager::run(llvm::Module&) /home/pferro/repos/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1649:1
#18 0x0000000004a37b51 compileModule(char**, llvm::LLVMContext&) /home/pferro/repos/llvm-project/llvm/tools/llc/llc.cpp:753:34
#19 0x0000000004a353f1 main /home/pferro/repos/llvm-project/llvm/tools/llc/llc.cpp:411:35
#20 0x00007f0dd0129248 __libc_start_call_main (/lib64/libc.so.6+0x3248)
#21 0x00007f0dd012930b __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x330b)
#22 0x0000000004a33fa5 _start (/home/pferro/repos/llvm-project/llvm/build-Debug/bin/llc+0x4a33fa5)
[1]    2423725 IOT instruction (core dumped)  llc -mtriple=nvptx64-nvidia-cuda -O3 -verify-machineinstrs <
```

Artem-B replied:
> shouldn't we want to check that optimizations may not introduce new instructions

There are two things -- optimization results must be valid. And that must be tested.
Back-end must handle valid inputs.
Those are somewhat independent things. 

The crash above is a bug (please file an issue for it, if one does not exist yet) and needs to be addressed.
whatever fixes it should also include the test that the code produces correct results regardless of optimization level.

This patch should concentrate on generating valid v2f32 operations where they are supported, and split them into scalar ops on older GPUs, and all of that should work about the same regardless of optimizations.

Prince781 replied:
> The crash above is a bug (please file an issue for it, if one does not exist yet) and needs to be addressed.
whatever fixes it should also include the test that the code produces correct results regardless of optimization level.

Okay, I think I understand: since this is an issue with DAGCombiner/initial DAG lowering, we shouldn't have a test in `CodeGen` for it.

Artem-B replied:
We should have a test, just not somewhere in the code that happens to use the broken output as input.

I'll need to take a look at the failure above and see why it ends up being optimization-dependent, and if it is supposed to be affected by optimizations, why we end up generating something invalid.

Artem-B commented on llvm/test/CodeGen/NVPTX/f32x2-instructions.ll:
```diff
@@ -0,0 +1,2665 @@
+; CHECK-O0-EMPTY:
+; CHECK-O0-NEXT:  // %bb.0:
+; CHECK-O0-NEXT:    ld.param.v2.f32 {%f1, %f2}, [test_fadd_param_0];
+; CHECK-O0-NEXT:    mov.b32 %r1, %f1;
+; CHECK-O0-NEXT:    cvt.u64.u32 %rd2, %r1;
+; CHECK-O0-NEXT:    mov.b32 %r2, %f2;
+; CHECK-O0-NEXT:    cvt.u64.u32 %rd3, %r2;
+; CHECK-O0-NEXT:    shl.b64 %rd4, %rd3, 32;
+; CHECK-O0-NEXT:    or.b64 %rd5, %rd2, %rd4;
```
This looks rather bad. We should probably custom-lower v2f32 loads/stores similar to how we load v2f16/v4i8 using a larger integer type.

Prince781 commented:
Hi,

I've updated the patch to legalize `v2f32` as `i64` and plumbed through the various areas. I've also introduced DAGCombine rules to ensure we still codegen `ld.vN` / `st.vN` when it makes sense to do so (when all accesses are of elements of `v2f32`). Along with some other rules for bitwise operations, this results in PTX codegen looking roughly the same if you're not using `.f32x2` instructions.

The patch probably needs more coverage and I'd also appreciate in suggestions in that direction.

Prince781 commented:
Pinging reviewers: @AlexMaclean, @Artem-B, @arsenm, @justinfargnoli, @durga4github

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -5051,26 +5080,395 @@ PerformFADDCombineWithOperands(SDNode *N, SDValue N0, SDValue N1,
+static SDValue PerformLoadCombine(SDNode *N,
+                                  TargetLowering::DAGCombinerInfo &DCI) {
+  auto *MemN = cast<MemSDNode>(N);
+  // only operate on vectors of f32s / i64s
```
Why is this?

Prince781 replied:
The pattern is only intended for rewriting `v2f32 = load` -> `f32,f32 = load`, according to the comments I've now added.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -5051,26 +5080,395 @@ PerformFADDCombineWithOperands(SDNode *N, SDValue N0, SDValue N1,
+  return {{NewLD, LoadChain}};
+}
+
+static SDValue PerformLoadCombine(SDNode *N,
```
Could you add a comment here explaining what this function is attempting to accomplish?

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -5681,6 +6079,150 @@ static SDValue combineADDRSPACECAST(SDNode *N,
+  EVT NarrowVT = N->getValueType(0);
+  EVT WideVT = Op.getValueType();
+
+  // v2[b]f16 = fp_round (v2f32 A)
+  // -> v2[b]f16 = (build_vector ([b]f16 = fp_round (extractelt A, 0)),
+  //                             ([b]f16 = fp_round (extractelt A, 1)))
```
Why is this required?

Prince781 replied:
Added a comment explaining. `v2f16 = fp_round v2f32` is a legal operation but with two f32 input registers, not a single `v2f32`. So we have to expand this node. A later pattern match will turn it into `cvt.f16x2.f32`.

AlexMaclean reviewed (commented):
Left some comments but will need to review more thoroughly. This change is quite large and complex, would it be possible to break it up in any way? Such as adding some of the DAG combine rules in a separate change?

Prince781 committed:
[NVPTX] Support f32x2 instructions for sm_100+

Lower `fadd`, `fsub`, `fmul`, and `fma` to f32x2 variants introduced in
PTX 8.6 for sm_100+. Adds a new register class for v2f32 as a b64
register in PTX. This causes other vector operations like loads and
stores to lower as .b64 instead of .v2.b32 as appropriate.

Also update test cases to use the autogenerator.

Prince781 commented:
**Update:**

I've streamlined the PR and it's much smaller now. Many of the changes in the test cases are just register renames.

The structuring/destructuring `mov.b64`s may look ugly, but they're mostly NOPs at the SASS level because 64-bit registers in PTX are emulated with 2 32-bit registers in SASS. I prefer not to introduce too many changes to existing pre-SM100 PTX, and there are only a few instructions supporting `f32x2` currently, hence why I'm forcing breaking apart loads/stores of v2f32. 

I have some ideas for cleaning up these `mov`s which I'd like to present in another PR, just to keep this one easy to review.

Prince781 commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -3595,6 +3635,9 @@ NVPTXTargetLowering::LowerReturn(SDValue Chain, CallingConv::ID CallConv,
   unsigned I = 0;
   for (const unsigned NumElts : VectorInfo) {
+    // amount to subdivide. If not v2f32, we don't consider packing
+    const unsigned PackingAmt = VTs[I] == MVT::v2f32 ? 2 : 1;
```
This change allows us to keep generating `st.param.v2.b32` for `ret <2 x float>` instead of `st.param.b64`, because I want to keep pre-SM100 code intact. It may make sense to do this for `v2f16` / `v2bf16` as well, but I'd rather explore that in another PR.

Prince781 replied:
I think I can pull this out and the change in `LowerFormalArguments` into a separate PR.

Prince781 commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -903,6 +921,11 @@ NVPTXTargetLowering::NVPTXTargetLowering(const NVPTXTargetMachine &TM,
+  // Expand v2f32 = fp_extend
+  setOperationAction(ISD::FP_EXTEND, MVT::v2f32, Expand);
+  // Expand v2[b]f16 = fp_round v2f32
+  setOperationAction(ISD::FP_ROUND, {MVT::v2bf16, MVT::v2f16}, Expand);
```
This wasn't needed before because `fp_round v2f32` was already being expanded since `v2f32` was an invalid type. Now we need it, otherwise this node will persist to isel.

Prince781 commented on llvm/test/CodeGen/NVPTX/f16x2-instructions.ll:
```diff
@@ -1943,10 +1954,12 @@ define <2 x half> @test_copysign_f32(<2 x half> %a, <2 x float> %b) #0 {
-; CHECK-F16-NEXT:    ld.param.v2.b32 {%r2, %r3}, [test_copysign_f32_param_1];
 ; CHECK-F16-NEXT:    ld.param.b32 %r1, [test_copysign_f32_param_0];
+; CHECK-F16-NEXT:    ld.param.v2.b32 {%r2, %r3}, [test_copysign_f32_param_1];
+; CHECK-F16-NEXT:    mov.b64 %rd1, {%r2, %r3};
```
I've found out this is due to leftover `CopyToReg` nodes that are optimized away in `-O3` but this test case is in `-O0`.

Prince781 commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -5103,6 +5158,54 @@ static SDValue PerformFADDCombine(SDNode *N,
+                                       TargetLowering::DAGCombinerInfo &DCI) {
+  // Convert (fop.x2 (vector_shuffle<1,u> V), V) -> ((fop V:1, V:0), undef)
+  const EVT VectorVT = N->getValueType(0);
+  if (!(VectorVT == MVT::v2f32 || Isv2x16VT(VectorVT)))
```
Sure. https://github.com/llvm/llvm-project/pull/143943

Prince781 commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -3126,22 +3155,38 @@ SDValue NVPTXTargetLowering::LowerVASTART(SDValue Op, SelectionDAG &DAG) const {
-      std::tie(Ops[0], Ops[1]) = expandUnalignedLoad(Load, DAG);
-      return DAG.getMergeValues(Ops, SDLoc(Op));
-    }
+  if (!(Isv2x16VT(VT) || VT == MVT::v4i8 || VT == MVT::v2f32))
```
Done.

Prince781 commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -1245,7 +1251,9 @@ bool NVPTXDAGToDAGISel::tryLDGLDU(SDNode *N) {
     // vectors of 8/16bits type are loaded/stored as multiples of v4i8/v2x16
     // elements.
-    if ((EltVT == MVT::f16 && OrigType == MVT::v2f16) ||
+    // Packed vector types are loaded/stored in a single register.
```
This was removed in the rebase.

Prince781 commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -471,8 +471,13 @@ bool NVPTXDAGToDAGISel::tryEXTRACT_VECTOR_ELEMENT(SDNode *N) {
   MVT VT = Vector.getSimpleValueType();
-  if (!Isv2x16VT(VT))
-    return false;
+  auto Opcode = NVPTX::I32toV2I16;
```
Sure.

Prince781 commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -942,12 +965,14 @@ NVPTXTargetLowering::NVPTXTargetLowering(const NVPTXTargetMachine &TM,
     setOperationAction(Op, MVT::f64, Legal);
     setOperationAction(Op, MVT::v2f16, Expand);
     setOperationAction(Op, MVT::v2bf16, Expand);
+    setOperationAction(Op, MVT::v2f32, Expand);
```
Done.

AlexMaclean reviewed (commented):
Nice. I think this is getting pretty close.

Prince781 commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -5098,7 +5148,7 @@ static SDValue combinePackingMovIntoStore(SDNode *N,
   EVT ElementVT = N->getOperand(Front).getValueType();
 
-  if (!Isv2x16VT(ElementVT))
+  if (!isPackedVectorTy(ElementVT))
```
We want to bail out early if the elements being stored are legal non-packed vector types (i32, etc).

As for v4i8, I realize this won't get triggered because in the case of something like:

```mir
v: v4i8 = BUILD_VECTOR a, b, c, d
STORE v
```
The `BUILD_VECTOR` will be turned into a PRMT after legalization but before we hit this combiner rule. Rather than having `!isPackedVectorVy(ElementVT) || ElementVT == MVT::v4i8` I think it's better to leave this check as-is since it's not harmful and keeps the code looking cleaner.

Prince781 commented on llvm/test/CodeGen/NVPTX/f32x2-instructions.ll:
```diff
@@ -0,0 +1,1699 @@
+; CHECK-EMPTY:
+; CHECK-NEXT:  // %bb.0:
+; CHECK-NEXT:    ld.param.v2.b64 {%rd1, %rd2}, [test_fadd_imm_0_v4_ftz_param_0];
+; CHECK-NEXT:    mov.b32 %r1, 0f40800000;
+; CHECK-NEXT:    mov.b32 %r2, 0f40400000;
+; CHECK-NEXT:    mov.b64 %rd3, {%r2, %r1};
```
At the SASS level, `%rd3` just becomes an alias for 32-bit registers `%r1` and `%r2` so the last mov is a NOP.

AlexMaclean reviewed (approved):
LGTM. Given this is a fairly large change please wait for additional review prior to landing.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXUtilities.h:
```diff
@@ -85,8 +85,14 @@ inline unsigned promoteScalarArgumentSize(unsigned size) {
+          VT == MVT::v2i16 || VT == MVT::v2f32);
+}
+
+inline bool isPackedElementTy(EVT VT) {
```
What makes these types 'packed' ? The function could use a comment explaining what it means.
It clearly not the common meaning of packed as related to aligning struct fields w/o alignment gaps.

Does it mean that it's one of the types for which we have vectorized PTX instructions? It could use a more distinct name.

Prince781 replied:
"Packed" here refers to whether a single vector type or two or more element types can fit into a single register.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -330,22 +329,30 @@ static void ComputePTXValueVTs(const TargetLowering &TLI, const DataLayout &DL,
+        // v2i8 is promoted to v2i16
+        NumElts = 1;
+        EltVT = MVT::v2i8;
+      } else if (isPackedElementTy(EltVT) && NumElts % 2 == 0 &&
+                 isPowerOf2_32(NumElts)) {
```
This looks like a rather roundabout way to check what we want. The condition is probably true for just a handful of types, so it may be more readable to just enumerate them, rather than relying on constraining the parameter space to carve out just those types.

Prince781 replied:
Okay, I redid this section to rely on a declared list of supported types (see changes in `NVPTXUtilities.h`) and added more comments.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2069,7 +2091,7 @@ SDValue NVPTXTargetLowering::LowerBITCAST(SDValue Op, SelectionDAG &DAG) const {
                                                SelectionDAG &DAG) const {
   EVT VT = Op->getValueType(0);
-  if (!(Isv2x16VT(VT) || VT == MVT::v4i8))
+  if (!isPackedVectorTy(VT) || !VT.is32BitVector())
```
Nit: Rephrased as `if (!(isPackedVectorTy(VT) && VT.is32BitVector()))` would be easier to read, IMO.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2119,15 +2141,12 @@ SDValue NVPTXTargetLowering::LowerBUILD_VECTOR(SDValue Op,
-    llvm_unreachable("Unsupported type");
-  }
+
+  APInt Value(32, 0);
+  const unsigned NumElements = VT.getVectorNumElements();
+  const unsigned ShiftAmount = 32 / NumElements;
+  for (unsigned ElementNo : seq(NumElements))
+    Value |= GetOperand(Op, ElementNo).shl(ElementNo * ShiftAmount);
```
This could use a comment, as it's not as obvious what we're doing here as the old code.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -3064,22 +3084,27 @@ SDValue NVPTXTargetLowering::LowerVASTART(SDValue Op, SelectionDAG &DAG) const {
-      std::tie(Ops[0], Ops[1]) = expandUnalignedLoad(Load, DAG);
-      return DAG.getMergeValues(Ops, SDLoc(Op));
-    }
+  if (!isPackedVectorTy(VT))
+    return SDValue();
+
+  // v2f32/v2f16/v2bf16/v2i16/v4i8 are legal, so we can't rely on legalizer to
+  // handle unaligned loads and have to handle it here.
+  LoadSDNode *Load = cast<LoadSDNode>(Op);
+  EVT MemVT = Load->getMemoryVT();
+  if (!allowsMemoryAccessForAlignment(*DAG.getContext(), DAG.getDataLayout(),
+                                      MemVT, *Load->getMemOperand())) {
+    SDValue Ops[2];
+    std::tie(Ops[0], Ops[1]) = expandUnalignedLoad(Load, DAG);
+    return DAG.getMergeValues(Ops, SDLoc(Op));
   }
 
   return SDValue();
```
Early return just clutters things here, as we jo just one thing, and still have unhandled cases returning SDValue() below.

I'd just put the processing under `if (isPackedVectorTy(VT))`.

Artem-B commented on llvm/test/CodeGen/NVPTX/f32x2-instructions.ll:
```diff
@@ -0,0 +1,1962 @@
+; RUN: %}
+; RUN: llc < %s -mcpu=sm_100 -O0 -disable-post-ra -frame-pointer=all           \
+; RUN: -verify-machineinstrs | FileCheck --check-prefixes=CHECK,CHECK-F32X2 %s
+; RUN: %if ptxas %{                                                            \
```
This should use if `ptxas-x.y`, otherwise the tests will fail if used with older PTXAS that does not support the new instructions.

Artem-B reviewed (approved):
Few nits. LGTM overall.

Artem-B reviewed (approved):
LGTM.

rupprecht commented:
> It appears that the patch may have started triggering an assertion somewhere in `MachineLICM`: https://godbolt.org/z/18hssYfME
> 
> @rupprecht how much time do we have to figure out if it's something we can fix forward?

Sometime in the next couple of days should be OK. Beyond that, i.e. next week, it might be better to go for a revert.

@bangtianliu, I notice you reverted this in iree. Do you have more details on what kind of failure you saw? Was it anything like https://godbolt.org/z/18hssYfME?

bangtianliu commented:
> > It appears that the patch may have started triggering an assertion somewhere in `MachineLICM`: https://godbolt.org/z/18hssYfME
> > @rupprecht how much time do we have to figure out if it's something we can fix forward?
> 
> Sometime in the next couple of days should be OK. Beyond that, i.e. next week, it might be better to go for a revert.
> 
> @bangtianliu, I notice you reverted this in iree. Do you have more details on what kind of failure you saw? Was it anything like https://godbolt.org/z/18hssYfME?

Here is the link to the CI error: https://github.com/iree-org/iree/actions/runs/16273907625/job/45959015129?pr=21364#step:6:19287 in IREE when I was trying to do LLVM integration.

rupprecht commented:
> > > It appears that the patch may have started triggering an assertion somewhere in `MachineLICM`: https://godbolt.org/z/18hssYfME
> > > @rupprecht how much time do we have to figure out if it's something we can fix forward?
> > 
> > 
> > Sometime in the next couple of days should be OK. Beyond that, i.e. next week, it might be better to go for a revert.
> > @bangtianliu, I notice you reverted this in iree. Do you have more details on what kind of failure you saw? Was it anything like https://godbolt.org/z/18hssYfME?
> 
> Here is the link to the CI error: https://github.com/iree-org/iree/actions/runs/16273907625/job/45959015129?pr=21364#step:6:19287 in IREE when I was trying to do LLVM integration.

Thanks for the pointer! Yep, looks like the same error:

```
FAILED: tests/e2e/matmul/e2e_matmul_cuda_f32_cuda_cuda_matmul.vmfb /__w/iree/iree/build/tests/e2e/matmul/e2e_matmul_cuda_f32_cuda_cuda_matmul.vmfb 
cd /__w/iree/iree/build/tests/e2e/matmul && /__w/iree/iree/build/tools/iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=cuda --iree-cuda-target=sm_80 /__w/iree/iree/build/tests/e2e/matmul/e2e_matmul_cuda_f32_cuda_cuda_matmul.mlir -o /__w/iree/iree/build/tests/e2e/matmul/e2e_matmul_cuda_f32_cuda_cuda_matmul.vmfb --iree-hal-executable-object-search-path=\"/__w/iree/iree/build\"
iree-compile: /__w/iree/iree/third_party/llvm-project/llvm/lib/CodeGen/MachineLoopInfo.cpp:281: bool llvm::MachineLoop::isLoopInvariant(llvm::MachineInstr &, const llvm::Register) const: Assertion `MRI->getVRegDef(Reg) && "Machine instr not mapped for this vreg?!"' failed.
Please report issues to https://github.com/iree-org/iree/issues and include the crash backtrace.
Stack dump:
0.	Running pass 'Function Pass Manager' on module 'e2e_matmul_cuda_f32_cuda_cuda_matmul_linked'.
1.	Running pass 'Early Machine Loop Invariant Code Motion' on function '@matmul_512x1000xf32_times_1000x4xf32_into_512x4xf32_dispatch_0_matmul_512x4x1000_f32'
```

Artem-B commented:
Here's where we losing that load :

```
Legalizing: t6: ch = CopyToReg t0, Register:v2f32 %0, t4
Legal node: nothing to do

Combining: t6: ch = CopyToReg t0, Register:v2f32 %0, t4

Legalizing: t4: v2f32,ch = load<(dereferenceable invariant load (s64), addrspace 101)> t0, ExternalSymbol:i64'Equal_GPU_DT_COMPLEX64_DT_BOOL_kernel_param_0', undef:i64
Legalizing non-extending load operation

Combining: t4: v2f32,ch = load<(dereferenceable invariant load (s64), addrspace 101)> t0, ExternalSymbol:i64'Equal_GPU_DT_COMPLEX64_DT_BOOL_kernel_param_0', undef:i64
Creating constant: t7: i64 = Constant<0>
Creating new node: t8: f32,f32,ch = NVPTXISD::LoadV2<(dereferenceable invariant load (s64), addrspace 101)> t0, ExternalSymbol:i64'Equal_GPU_DT_COMPLEX64_DT_BOOL_kernel_param_0', undef:i64, Constant:i64<0>
Creating new node: t9: v2f32 = BUILD_VECTOR t8, t8:1

Replacing.1 t6: ch = CopyToReg t0, Register:v2f32 %0, t4

With: t0: ch,glue = EntryToken
 and 0 other values

```

Looks like we clobber `t6`, and from this point on the `LoadV2` is gone.

Prince781 commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -5022,23 +5075,28 @@ combineUnpackingMovIntoLoad(SDNode *N, TargetLowering::DAGCombinerInfo &DCI) {
 
+  // Remove dead CopyToReg nodes by folding them into the chain they reference
+  for (SDNode *CTR : DeadCopyToRegs)
+    DCI.CombineTo(CTR, CTR->getOperand(0));
```
This was just an optimization to improve the default PTX codegen. Without it, we have a lot of spurious movs in `-O0`, but these are eliminated in `-O3`. We can remove this optimization.

Artem-B replied:
I think we're just overzealous in considering all CopyToReg dead. Some of them may be legitimately needed.

The code that populates `DeadCopyToRegs` appears to stick all `CopyToReg` instructions there indiscriminately.

As a short-term fix, disabling the removal would do, but there may still be some benefit in eliminating redundant register copies, as long as we make sure that we do not lose the value we were copying.

Prince781 replied:
I think the real issue is CopyToReg may need to stay around due to use of the original register in other blocks. Whereas with no control flow, then it's dead for sure.

I agree on disabling this optimization for now until we have a better solution.

npanchen commented:
@Prince781 due to this change we have lots of runtime errors, i.e. accuracy mismatch which #149393 does not solve.
If @bangtianliu observed this too, it's worth to revert that change.

npanchen commented:
smallest reproducer:
```
define ptx_kernel void @test(float %0) {
  %2 = insertelement <2 x float> zeroinitializer, float %0, i64 0
  %3 = fptrunc <2 x float> %2 to <2 x bfloat>
  store <2 x bfloat> %3, ptr addrspace(1) null, align 2
  ret void
}
```

without the change isel generates:
```
bb.0 (%ir-block.1):
  %0:b32 = LD_i32 0, 0, 101, 3, 32, &test_param_0, 0 :: (dereferenceable invariant load (s32), addrspace 101)
  %1:b16 = CVT_bf16_f32 killed %0:b32, 5
  %3:b16 = IMPLICIT_DEF
  %2:b32 = V2I16toI32 killed %1:b16, killed %3:b16
  %4:b64 = IMOV64i 0
  ST_i32 killed %2:b32, 0, 0, 1, 16, killed %4:b64, 0 :: (store (s16) into `ptr addrspace(1) null`, addrspace 1)
  %5:b64 = IMOV64i 2
  ST_i32 0, 0, 0, 1, 16, killed %5:b64, 0 :: (store (s16) into `ptr addrspace(1) null` + 2, addrspace 1)
  Return
```

with the change:
```
bb.0 (%ir-block.1):
  %0:b32 = LD_i32 0, 0, 101, 3, 32, &test_param_0, 0 :: (dereferenceable invariant load (s32), addrspace 101)
  %1:b32 = FMOV32i float 0.000000e+00
  %2:b64 = V2I32toI64 killed %0:b32, killed %1:b32
  %3:b64 = IMOV64i 0
  ST_i64 %2:b64, 0, 0, 1, 16, killed %3:b64, 0 :: (store (s16) into `ptr addrspace(1) null`, addrspace 1)
  %4:b64 = SRLi64ri %2:b64, 16
  %5:b64 = IMOV64i 2
  ST_i64 killed %4:b64, 0, 0, 1, 16, killed %5:b64, 0 :: (store (s16) into `ptr addrspace(1) null` + 2, addrspace 1)
  %6:b64 = IMOV64i 6
  ST_i64 0, 0, 0, 1, 16, killed %6:b64, 0 :: (store (s16) into `ptr addrspace(1) null` + 6, addrspace 1)
  %7:b64 = IMOV64i 4
  ST_i64 0, 0, 0, 1, 16, killed %7:b64, 0 :: (store (s16) into `ptr addrspace(1) null` + 4, addrspace 1)
  Return
```

cc @Artem-B

npanchen commented:
> Looks like we need to add the following:
> 
> ```
>   setTruncStoreAction(MVT::v2f32, MVT::v2f16, Expand);
>   setTruncStoreAction(MVT::v2f32, MVT::v2bf16, Expand);
> ```
> 
> Otherwise the fp_round will get folded into the store which we can't handle.

Yes, I finally got a chance to debug it and realized there's incorrect handling of truncating stores for a long period of time. Will you(@AlexMaclean) or @Artem-B or @Prince781 create a fix ?

npanchen commented:
> Looks like we need to add the following:
> 
> ```
>   setTruncStoreAction(MVT::v2f32, MVT::v2f16, Expand);
>   setTruncStoreAction(MVT::v2f32, MVT::v2bf16, Expand);
> ```
> 
> Otherwise the fp_round will get folded into the store which we can't handle.

With both https://github.com/llvm/llvm-project/pull/149393 and https://github.com/llvm/llvm-project/pull/149571 my local testing is clean!

AlexMaclean commented:
@npanchen Would it be possible to attach the IR that the PTX was generated from?

npanchen commented:
@AlexMaclean @Prince781 ping

Prince781 commented:
@npanchen I compared the "good" and "bad" PTX and I notice there are now `mov.b64`s inside your async wgmma pipeline (between `wgmma.mma_async` calls). Register accesses inside are disallowed[^1] and ptxas will force `wgmma` to be synchronous.

[^1]: See [PTX ISA for `wgmma.mma_async`](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#asynchronous-warpgroup-level-matrix-instructions-wgmma-mma): "wgmma.fence instruction must be used to fence the register accesses of wgmma.mma_async instruction from their prior accesses. Otherwise, the behavior is undefined."

Prince781 commented:
@npanchen if the original source was CUDA C++, then you can use the [same trick CUTLASS uses](https://github.com/NVIDIA/cutlass/blob/main/include/cute/arch/mma_sm90_gmma.hpp#L86-L95) where you pass the operand through a `asm volatile` to get the desired anti-dependency with your `wgmma.mma_async` inline ASM call.

Prince781 commented:
@npanchen it looks like there's a similar thing at the LLVM IR level ([`@llvm.arithmetic.fence`](https://llvm.org/docs/LangRef.html#llvm-arithmetic-fence-intrinsic)) you could also look into, although I haven't tried it myself.

Prince781 commented:
> out or it's simply expected that mem2reg will happen, i.e. it won't work with O0 correctly too.

Yes, this is actually necessary to get rid of the intermediate registers. However, it will still work correctly in O0, it will just be serialized by the compiler (ptxas) as a fail-safe, so you'll lose a lot of perf.

> btw, another concern I got from internal developers is that this PR potentially will impact perf badly for sm_90[a]. At least SASS does have more instructions (with more fences that will be even more). Is there a plan to limit new packing mechanism to only use cases mentioned in description and the sm_100+ target ?

Perhaps we can improve things pre-sm_100 with more DAGCombiner rules. I can take a closer look at your example later. If you have any more please post them here. If you're talking about using arithmetic fences, those shouldn't compile to anything in SASS.

Artem-B commented:
@Prince781 it appears that the PTX generated after this patch (and subsequent fixes) triggers a miscompilation in ptxas.

So far it shows up as an invalid result of a fairly large computation on one test that resists being reduced, but the issue goes away if the PTX is compiled with lower ptxas optimization settings. The issue shows up with both cuda 12.8.1 and 12.9.1.

Attached is the IR and PTX files for the working/failing kernel, but I have no test usable for reducing the kernel further.
[ptxas-miscompile-reproducer.zip](https://github.com/user-attachments/files/21495545/ptxas-miscompile-reproducer.zip)

@AlexMaclean, @gonzalobg  : ^^^ FYI.

Prince781 commented:
@Artem-B I've forwarded this to our internal ptxas team btw.

Artem-B commented:
It may also be related to https://github.com/llvm/llvm-project/issues/151580. This miscompilation, the performance regression in #151580, and another crashing kernel which we can't yet root cause were bisected to this patch. I was unable to find anything wrong with the PTX (other than few issues we've already fixed post this patch). What looks like the expected benign changes in our use of the registers, largely amounting to extra b32/b64 splits/merges seem to hit some unusual code paths in ptxas.

akuegel commented:
@Prince781 We have a reproducer for a crash when lowering to PTX. I have attached it here. It works fine when v2f32 is disabled (i.e. without the changes from this PR). Can you please take a look?
[reproducer.ll.txt](https://github.com/user-attachments/files/21655779/reproducer.ll.txt)

Prince781 commented:
@alexreinking have you repro'd with LLVM trunk? This may have been fixed.

Prince781 commented:
@alexreinking FYI this may have already been fixed in https://github.com/llvm/llvm-project/commit/d63ab5467dcae0492e2f4def336ddbb73ce10dc5

Artem-B commented:
Let's see if the patch can be cherry-picked into the next minor release https://github.com/llvm/llvm-project/pull/149393#issuecomment-3229424628

Prince781 commented:
@Artem-B Our PTXAS team is taking a look at this now, and wondering if there's a smaller or runnable repro. Thanks!

Artem-B commented:
Unfortunately I do not have anything better than what I've reported earlier.
